app.ts

```diff
@@ -0,0 +1,140 @@
+
+        // calculate row number based on table header + row heights + borders in CSS and ensure its never less than 1
+        let numRows = Math.floor((contentHeight - CONTROLS_HEIGHT - 26) / 46);
+        numRows = numRows <= 0 ? 1 : numRows;
+
+        // set number of records to get from server and ensure that the last record doesnt exceed the limit.
+        let endRecord = (firstRecord + numRows - 1);
+        endRecord = (endRecord >= RECORD_COUNT ? RECORD_COUNT - 1 : endRecord);
+
+        // get data from server
+        const columnsString = getData("http://localhost:2050/columns");
+        let recordsText = getData("http://localhost:2050/records?from=" + firstRecord + "&to=" + endRecord);
+
+        // parse text from api call into arrays
+        const columnsArray = parseText(columnsString);
+        let recordsTable = parseText(recordsText);
+
+        // write html for table header
+        for (const i of columnsArray) {
+            let th = document.createElement('th');
+            this.myTable.appendChild(th);
+
+            th.appendChild(document.createTextNode(" " + i));
+        }
+
+        // write html for table
+        for (const iEntry of recordsTable) {
+
+            let tr = document.createElement('tr');
+            this.myTable.appendChild(tr);
+
+            for (const jEntry of iEntry) {
+
+                let td = document.createElement('td');
+                td.appendChild(document.createTextNode(" " + jEntry));
+                tr.appendChild(td);
+            }
+        }
+    }
+}
+
+function createPage(firstRecord) {
+
+    // ensure that the user input is a number and within range of available records
+    if (firstRecord >= RECORD_COUNT || firstRecord < 0 || isNaN(firstRecord)) {
+        alert('please select a NUMBER between 0 and ' + (RECORD_COUNT - 1));
+        return;
+    }
+
+    let contentsDiv = this.content;
+    contentsDiv.innerHTML = ''; // empty table div to fill in a new table
+    let tableElement = new tableGenerator(contentsDiv, parseInt(firstRecord)); // create table using the tableGenerator class
+
+    // keep the ID of the first and last table entries to use for next and previous page buttons
+    LAST_ENTRY = contentsDiv.lastChild.lastChild.firstChild.textContent;
+    CURRENT_ENTRY = firstRecord;
+};
+
+// fucntion used by the next page buttton, requires the last entry from previously generated table
+function nextPage() {
+    createPage(parseInt(LAST_ENTRY) + 1);
+}
+
+// fucntion used by the previous page buttton, requires the first and last entry from previously generated table
+function prevPage() {
+
+    let newfirstEntry = CURRENT_ENTRY - (LAST_ENTRY - CURRENT_ENTRY) - 1;
+
+    // if the new first entry is less than zero, take user to the first page 
+    newfirstEntry <= 0 ? createPage(0) : createPage(newfirstEntry);
+}
+
+// hint for the user
+function helpFunc() {
+    alert('Table for viewing records with IDs 0 to '
```

Comment:
IMO a better way to do multi-line (or longer) strings is with backticka as follows
```suggestion
    alert(`
        Table for viewing records with IDs 0 to ${RECORD_COUNT - 1}  (Numbers Only)
        To 
```

(Not sure why the 'with' is highlighted in orange 😕